### ISSUE TITLE
Int8 quantization for microcontroller

### ISSUE THREAD

**Issue by aqibsaeed** [larq/compute-engine#736]
Hi,

Is there a way to get a int8 quantized model that can eventually run on a microcontroller? I am converting a binary densenet keras model (https://docs.larq.dev/zoo/api/literature/#binarydensenet28) as follows but it does not result in a quantized model. Am I missing something here? 

```
tflite_model = lce.convert_keras_model(kera_model, 
    inference_input_type=tf.int8, 
    inference_output_type=tf.int8,
    experimental_default_int8_range=(-3,3))
```

Thanks in advance.

**Comment by aqibsaeed** [comment]
Any pointers on how to solve this?

**Comment by CNugteren** [comment]
How did you conclude that it does not result in a quantized model? I got a INT8/BNN model by running the code snippet below:
```python
from pathlib import Path
import larq_zoo
import larq_compute_engine as lce
import tensorflow as tf

keras_model = larq_zoo.literature.BinaryDenseNet28(
    input_shape=None,
    input_tensor=None,
    weights="imagenet",
    include_top=True,
    num_classes=1000
)

tflite_model = lce.convert_keras_model(keras_model,
    inference_input_type=tf.int8, 
    inference_output_type=tf.int8,
    experimental_default_int8_range=(-3,3)
)

Path("model.tflite").write_bytes(tflite_model)
```
I inspected the resulting `.tflite` file in [Netron](https://netron.app/).

**Comment by aqibsaeed** [comment]
I tried the following:

```
model_a = lqz.literature.BinaryDenseNet28(
    input_shape=(32,32,3),
    weights=None,
    include_top=True,
    num_classes=10
)

with lq.context.quantized_scope(True):
  weights = model_a.get_weights()
  model_a.set_weights(weights)


tflite_model = lqce.convert_keras_model(model_a,
    inference_input_type=tf.int8, 
    inference_output_type=tf.int8,
    experimental_default_int8_range=(-3,3)
)

Path("model.tflite").write_bytes(tflite_model)
```



Now if I submit it to `https://plumerai.com/benchmark`: I get the following response back: `Your model contains layers that aren't INT8 quantized but instead use floating-point values. This is not suitable for microcontrollers. You can find information on how to quantize your model in the TensorFlow documentation.`

**Comment by aqibsaeed** [comment]
I do not get the above mentioned error when I use your snippet. Maybe I am missing something here `experimental_default_int8_range`.

**Comment by CNugteren** [comment]
Indeed, if I run your snippet I get a model that has a float layer in there. You can see this in Netron:
![float](https://user-images.githubusercontent.com/1162108/167377072-759d4ee6-8f8c-4d88-8ec1-7ad10b52e929.png)

@Tombana , do you perhaps have any idea why this might happen?

**Comment by aqibsaeed** [comment]
There's also a difference at the start 

Custom CIFAR-10 Model

![image](https://user-images.githubusercontent.com/5706772/167379287-d966f5b7-f58c-4e45-aeb2-c62ec4a58818.png)

IN-1K Model

![image](https://user-images.githubusercontent.com/5706772/167379372-8490b014-88f3-4c9d-9873-5d4ca1bf4f1c.png)

**Comment by Tombana** [comment]
I'm not sure, this sounds like a bug in the converter, it might be in the tensorflow converter itself. From what I see in the two code snippets, the difference is only in

```python
with lq.context.quantized_scope(True):
  weights = model_a.get_weights()
  model_a.set_weights(weights)
```

That shouldn't affect the outcome of the converter though.

Which version of `larq-compute-engine` was used for this?

**Comment by aqibsaeed** [comment]
I tested converting removing this

```
with lq.context.quantized_scope(True):
  weights = model_a.get_weights()
  model_a.set_weights(weights)
```

but it does not work.

`flatbuffers-1.12 larq-compute-engine-0.7.0`

**Comment by aqibsaeed** [comment]
I think the problem occurs if one specify `input_shape` or `input_tensor` when creating model object `lqz.literature.BinaryDenseNet28`.

**Comment by CNugteren** [comment]
I managed to reproduce the issue with a minimal example without the zoo, without binary layers, with just 3 layers:
```python
from pathlib import Path
import larq_compute_engine as lce
import tensorflow as tf

image_input = tf.keras.layers.Input(shape=(32, 32, 3))
x = image_input
x = tf.keras.layers.Conv2D(64, kernel_size=3)(x)
x = tf.keras.layers.MaxPool2D(3)(x)
x = tf.keras.layers.BatchNormalization()(x)
keras_model = tf.keras.Model(inputs=image_input, outputs=x)

tflite = lce.convert_keras_model(
    keras_model,
    inference_input_type=tf.int8,
    inference_output_type=tf.int8,
    experimental_default_int8_range=(-3, 3),
)
Path("model.tflite").write_bytes(tflite)
```
With `larq-compute-engine==0.7.0 tensorflow==2.8.0 keras==2.8.0` this gives:
![image](https://user-images.githubusercontent.com/1162108/167869403-edde613d-9066-4c66-855c-085eb4e53baf.png)

If I remove any of the three layers (`Conv2D`, `MaxPool2D`, or `BatchNormalization`) then it does produce the proper int8 model.

If I downgrade to `larq-compute-engine==0.6.2 tensorflow==2.6.1 keras==2.6.0` then it looks better but still the first Conv2D layer is in float:
![image](https://user-images.githubusercontent.com/1162108/167870703-386df2c0-e491-4d6d-8d38-bb8c385a0d57.png)

If I instead of LCE use the standard TFLite post-training quantizer then everything becomes INT8 as expected:
```python
def representative_dataset():
    for _ in range(10):
      data = np.random.rand(1, 32, 32, 3)
      yield [data.astype(np.float32)]

converter = tf.lite.TFLiteConverter.from_saved_model("test_keras_model")
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.int8  # or tf.uint8
converter.inference_output_type = tf.int8  # or tf.uint8
tflite = converter.convert()
Path("model.tflite").write_bytes(tflite)
```

So in conclusion this seems to be an LCE issue, and we will investigate further.

**Comment by aqibsaeed** [comment]
Interesting. Thank you for looking into it.

**Comment by aqibsaeed** [comment]
Got it! I just double checked if I load model weights, conversion works fine. Thanks again for looking into this. Closing this now.
